docs/blog/2019-03-03-new-schema-customization/index.md

```diff
@@ -0,0 +1,270 @@
+
+```graphql
+# For this type `name` won't be added
+@dontInfer()
+type AuthorJson implements Node {
+  birthday: Date
+}
+
+# For this type `name` won't be added but birtday won't get a Gatsby resolver for date formatting
+@dontInfer(noDefaultResolvers: true)
+type AuthorJson implements Node {
+  birthday: Date
+}
+
+
+# For this type all fields `name`. Current default behaviour, but allows one to be explicit about it.
+@infer()
+type AuthorJson implements Node {
+  id: ID!
+}
+
+# Birthday will be Date, but we won't add a Gatsby resolver for date formatting
+@infer(noDefaultResolvers: true)
+type AuthorJson implements Node {
+  id: ID!
+}
+```
+
+## `createResolvers`
+
+This is a similar API to `setFieldsOnNodeType` in that it allows to add new fields and resolvers for them. However, this one is ran last, so you'd have all the schema available to you in it. Also it's possible to extend types like `Query` to add more root resolvers.
+
+```js
+exports.createResolvers = ({ createResolvers, schema }) => {
+  createResolvers({
+    AuthorJson: {
+      // Modify birthday resolver so that it uses 1970-01-01 as default date
+      birthday: {
+        resolve(source, args, context, info) {
+          // original resolver available as "info.originalResolver"
+          if (Number.isNaN(new Date(source['birthday'])) {
+            return info.originalResolver({
+              ...source,
+              birthday: '1970-01-01',
+            }, args, context, info)
+          } else {
+            return info.originalResolver(source, args, context, info)
+          }
+        },
+      },
+    },
+  })
+}
+```
+
+It's also possible to create new root fields, for example one that will return all author names as strings.
+
+```js
+createResolvers({
+  Query: {
+    allAuthorFullNames: {
+      type: `[String!]!`,
+      resolve(source, args, context, info) {
+        const authors = context.nodeModel.getAllNodes(
+          {
+            type: `AuthorJson`,
+          },
+          { path: context.path }
+        )
+        return authors.map(author => author.name)
+      },
+    },
+  },
+})
+```
+
+Notice the `context.nodeModel`. We expose our internal node storage to the resolvers, so that one can fetch data from there. Full documentation of nodeModel is available at TODO.
+
+You can also see TODO ADD using resolvers example.
+
+# Why was it needed?
+
+The motivation to do this change is a two-fold one. We've seen many issues with automatically generated schemas. They can change from a minor data source modification, like changing one fields. The resulting schema will break queries on the pages and will leave users confused. Making inference smarter is just pouring more oil on already burning fire, because the core issue is not the inferrence, but lack of control. Therefore we wanted to give people control over the schema.
```

Comment:
```suggestion
The motivation to do this change is a two-fold one. We've seen many issues with automatically generated schemas. They can change from a minor data source modification, like changing one field. The resulting schema will break queries on the pages and will leave users confused. Making inference smarter is just pouring more oil on already burning fire, because the core issue is not the inference, but lack of control. Therefore we wanted to give people control over the schema.
```